File: webhook_analyzer.py
import json
from http.server import HTTPServer, BaseHTTPRequestHandler
from datetime import datetime
import sys

# ANSI Farben
class Colors:
    RESET = "\033[0m"
    RED = "\033[31m"
    GREEN = "\033[32m"
    YELLOW = "\033[33m"
    BLUE = "\033[34m"
    PURPLE = "\033[35m"
    CYAN = "\033[36m"
    WHITE = "\033[37m"

class WebhookHandler(BaseHTTPRequestHandler):
    def log_message(self, format, *args):
        pass  # Unterdrücke Standard-Logging

    def handle_request(self):
        timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")

        print(f"\n{Colors.CYAN}{'=' * 60}{Colors.RESET}")
        print(f"{Colors.GREEN}[{timestamp}]{Colors.RESET} Neue Anfrage empfangen")
        print(f"{Colors.CYAN}{'=' * 60}{Colors.RESET}")

        # Method und Path
        print(f"\n{Colors.YELLOW}Methode:{Colors.RESET}  {Colors.GREEN}{self.command}{Colors.RESET}")
        print(f"{Colors.YELLOW}Path:{Colors.RESET}     {Colors.WHITE}{self.path}{Colors.RESET}")
        print(f"{Colors.YELLOW}Von:{Colors.RESET}      {Colors.WHITE}{self.client_address[0]}:{self.client_address[1]}{Colors.RESET}")

        # Headers
        print(f"\n{Colors.PURPLE}--- Headers ---{Colors.RESET}")
        for key, value in sorted(self.headers.items()):
            print(f"{Colors.BLUE}{key}:{Colors.RESET} {value}")

        # Body lesen
        content_length = self.headers.get('Content-Length')
        if content_length:
            body = self.rfile.read(int(content_length))
            if body:
                print(f"\n{Colors.PURPLE}--- Body ({len(body)} Bytes) ---{Colors.RESET}")
                try:
                    json_data = json.loads(body.decode('utf-8'))
                    print(f"{Colors.WHITE}{json.dumps(json_data, indent=2, ensure_ascii=False)}{Colors.RESET}")
                except (json.JSONDecodeError, UnicodeDecodeError):
                    print(f"{Colors.WHITE}{body.decode('utf-8', errors='replace')}{Colors.RESET}")
        else:
            print(f"\n{Colors.PURPLE}--- Kein Body ---{Colors.RESET}")

        print(f"\n{Colors.CYAN}{'-' * 60}{Colors.RESET}")

        # Antwort senden
        self.send_response(200)
        self.send_header('Content-Type', 'application/json')
        self.end_headers()
        response = {
            "status": "received",
            "timestamp": timestamp,
            "message": "Webhook erfolgreich empfangen"
        }
        self.wfile.write(json.dumps(response).encode('utf-8'))

    def do_GET(self):
        self.handle_request()

    def do_POST(self):
        self.handle_request()

    def do_PUT(self):
        self.handle_request()

    def do_DELETE(self):
        self.handle_request()

    def do_PATCH(self):
        self.handle_request()

    def do_OPTIONS(self):
        self.handle_request()

def main():
    # Windows: ANSI Farben aktivieren
    if sys.platform == 'win32':
        import os
        os.system('')

    port = 9999
    server = HTTPServer(('0.0.0.0', port), WebhookHandler)

    print(f"{Colors.CYAN}========================================{Colors.RESET}")
    print(f"{Colors.GREEN}  Webhook Analyzer Server{Colors.RESET}")
    print(f"{Colors.CYAN}========================================{Colors.RESET}")
    print(f"{Colors.WHITE}Server lauscht auf Port {Colors.YELLOW}{port}{Colors.RESET}")
    print(f"{Colors.WHITE}URL: {Colors.YELLOW}http://localhost:{port}{Colors.RESET}")
    print(f"\n{Colors.PURPLE}Warte auf eingehende Webhooks...{Colors.RESET}")
    print(f"{Colors.WHITE}Druecke Ctrl+C zum Beenden{Colors.RESET}")
    print("-" * 50)

    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print(f"\n{Colors.YELLOW}Server beendet.{Colors.RESET}")
        server.shutdown()

if __name__ == "__main__":
    main()

File: webhook_analyzer.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"sort"
	"strings"
	"time"
)

const (
	colorReset  = "\033[0m"
	colorRed    = "\033[31m"
	colorGreen  = "\033[32m"
	colorYellow = "\033[33m"
	colorBlue   = "\033[34m"
	colorPurple = "\033[35m"
	colorCyan   = "\033[36m"
	colorWhite  = "\033[37m"
)

func main() {
	fmt.Println(colorCyan + "========================================" + colorReset)
	fmt.Println(colorGreen + "  Webhook Analyzer Server" + colorReset)
	fmt.Println(colorCyan + "========================================" + colorReset)
	fmt.Printf("%sServer lauscht auf Port %s9999%s\n", colorWhite, colorYellow, colorReset)
	fmt.Printf("%sURL: %shttp://localhost:9999%s\n\n", colorWhite, colorYellow, colorReset)
	fmt.Println(colorPurple + "Warte auf eingehende Webhooks..." + colorReset)
	fmt.Println(strings.Repeat("-", 50))

	http.HandleFunc("/", handleWebhook)

	err := http.ListenAndServe(":9999", nil)
	if err != nil {
		fmt.Printf("%sFehler: %s%s\n", colorRed, err.Error(), colorReset)
	}
}

func handleWebhook(w http.ResponseWriter, r *http.Request) {
	timestamp := time.Now().Format("2006-01-02 15:04:05")

	fmt.Printf("\n%s%s%s\n", colorCyan, strings.Repeat("=", 60), colorReset)
	fmt.Printf("%s[%s]%s Neue Anfrage empfangen\n", colorGreen, timestamp, colorReset)
	fmt.Printf("%s%s%s\n", colorCyan, strings.Repeat("=", 60), colorReset)

	// Method und Path
	fmt.Printf("\n%sMethode:%s  %s%s%s\n", colorYellow, colorReset, colorGreen, r.Method, colorReset)
	fmt.Printf("%sPath:%s     %s%s%s\n", colorYellow, colorReset, colorWhite, r.URL.Path, colorReset)

	if r.URL.RawQuery != "" {
		fmt.Printf("%sQuery:%s    %s%s%s\n", colorYellow, colorReset, colorWhite, r.URL.RawQuery, colorReset)
	}

	// Remote Address
	fmt.Printf("%sVon:%s      %s%s%s\n", colorYellow, colorReset, colorWhite, r.RemoteAddr, colorReset)

	// Headers
	fmt.Printf("\n%s--- Headers ---%s\n", colorPurple, colorReset)

	// Sortiere Headers für bessere Lesbarkeit
	var headerKeys []string
	for key := range r.Header {
		headerKeys = append(headerKeys, key)
	}
	sort.Strings(headerKeys)

	for _, key := range headerKeys {
		values := r.Header[key]
		fmt.Printf("%s%s:%s %s\n", colorBlue, key, colorReset, strings.Join(values, ", "))
	}

	// Body
	body, err := io.ReadAll(r.Body)
	if err != nil {
		fmt.Printf("%sFehler beim Lesen des Body: %s%s\n", colorRed, err.Error(), colorReset)
	} else if len(body) > 0 {
		fmt.Printf("\n%s--- Body (%d Bytes) ---%s\n", colorPurple, len(body), colorReset)

		// Versuche JSON zu formatieren
		var jsonData interface{}
		if err := json.Unmarshal(body, &jsonData); err == nil {
			prettyJSON, _ := json.MarshalIndent(jsonData, "", "  ")
			fmt.Printf("%s%s%s\n", colorWhite, string(prettyJSON), colorReset)
		} else {
			// Kein JSON, zeige raw
			fmt.Printf("%s%s%s\n", colorWhite, string(body), colorReset)
		}
	} else {
		fmt.Printf("\n%s--- Kein Body ---%s\n", colorPurple, colorReset)
	}

	fmt.Printf("\n%s%s%s\n", colorCyan, strings.Repeat("-", 60), colorReset)

	// Sende Antwort
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusOK)
	response := map[string]interface{}{
		"status":    "received",
		"timestamp": timestamp,
		"message":   "Webhook erfolgreich empfangen",
	}
	json.NewEncoder(w).Encode(response)
}
